irohad/ametsuchi/impl/storage_impl.cpp

```diff
@@ -737,88 +757,94 @@ namespace iroha {
                                 enable_prepared_transactions,
                                 prepared_block_name,
                                 std::move(log_manager))));
-                      } catch (const std::exception &e) {
-                        storage = expected::makeError(e.what());
-                      }
-                    },
-                    [&](const auto &error) { storage = error; });
-              },
-              [&](const auto &error) { storage = error; });
-      return storage;
+          } catch (const std::exception &e) {
+            return expected::makeError(e.what());
+          }
+        };
+      };
     }
 
-    boost::optional<std::unique_ptr<LedgerState>> StorageImpl::commit(
+    CommitResult StorageImpl::commit(
         std::unique_ptr<MutableStorage> mutable_storage) {
       auto storage = static_cast<MutableStorageImpl *>(mutable_storage.get());
 
       try {
         *(storage->sql_) << "COMMIT";
-        storage->committed = true;
-
-        storage->block_storage_->forEach(
-            [this](const auto &block) { this->storeBlock(block); });
-
-        return PostgresWsvQuery(*(storage->sql_),
-                                factory_,
-                                log_manager_->getChild("WsvQuery")->getLogger())
-                   .getPeers()
-            | [&storage](auto &&peers) {
-                return boost::optional<std::unique_ptr<LedgerState>>(
-                    std::make_unique<LedgerState>(std::move(peers),
-                                                  storage->getTopBlockHeight(),
-                                                  storage->getTopBlockHash()));
-              };
       } catch (std::exception &e) {
         storage->committed = false;
-        log_->warn("Mutable storage is not committed. Reason: {}", e.what());
-        return boost::none;
+        return expected::makeError(e.what());
+      }
+      storage->committed = true;
+
+      storage->block_storage_->forEach(
+          [this](const auto &block) { this->storeBlock(block); });
+
+      ledger_state_ = storage->getLedgerState();
+      if (ledger_state_) {
+        return expected::makeValue(ledger_state_.value());
+      } else {
+        // throw? because state in iroha is inconsistent with DB
```

Comment:
Do we have a politics about such cases? I remember we had a discussion but not sure about results?

Reply:
This kind of error is special, because we cannot recover from it and cannot continue (at least for now). So as I see we should either introduce different `Error` types (which is AFAIK planned in the future) or just use another error handling method (exception). Till we do not have the first option ready, I'm for the second.

Reply:
Again, summon @lebdron

Reply:
The problem with `throw` is that tests may not behave as expected - test thread can shut down ITF when it is between `forEach` and `getLedgerState`. I suggest to add a TODO instead to introduce different error types and handle them in callers as needed.